refactor: Use a dedicated HTTP client

dbellinghoven opened the pull request:
*Description of changes:*

Currently, every time an HTTP request is made, a new HTTP client is created. This makes the `Client` use a single HTTP client for all requests.

Signed-off-by: Dilan Bellinghoven <dbellinghoven@morningconsult.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

rliebz commented on lib/request.go:
```diff
@@ -19,19 +19,16 @@ type APIResponse struct {
 }
 
-func sendRequest(host, method, url, accessToken string, body interface{}, timeout int) (*APIResponse, error) {
+func (c *Client) sendRequest(host, method, url, accessToken string, body interface{}, _ int) (*APIResponse, error) {
```
Since this is a private method, and the timeout field passed is never referenced, can we change the signature here?

```go
func (c *Client) sendRequest(host, method, url, accessToken string, body interface{}) (*APIResponse, error) {
```

dbellinghoven replied:
Done

rliebz commented on lib/client.go:
```diff
@@ -651,6 +665,15 @@ func (c *Client) SetOptions(env string, timeout int) error {
+	client := c.Options.HTTPClient
+	if client == nil {
+		client = &http.Client{
+			Timeout: time.Second * time.Duration(timeout),
```
Took me a bit to wrap my head around the control flow here. I might throw an error if both `c.Options.HTTPClient` is set, and `timeout` is non-zero, because otherwise we'd be silently ignoring an option.

dbellinghoven replied:
Done

rliebz replied:
Actually I think we need to do this in the constructor as well.

dbellinghoven committed:
refactor: Use a dedicated HTTP client

Currently, every time an HTTP request is made, a new HTTP client is created.
This makes the `Client` use a single HTTP client for all requests.

Signed-off-by: Dilan Bellinghoven <dbellinghoven@morningconsult.com>
